### ISSUE TITLE
Can't enroll in webserver, TLS certification issues

### ISSUE THREAD

**Issue by zcatbear** [facebook/osquery#3303]
I'm having TLS certification issues when I try to enroll from a remote location. When I try to enroll the  system that the server is running on I have few issues, but when I try to enroll any other system, or even ping that system with a supplied certificate, I get a connection refused. 

When I just curl it on localhost I have no issues in the test_http_server.py if I do something as simple as:  
`curl https://localhost:8080 --insecure` because I get back `{"foo": "baz", "config": "baz"}` and can see that in the running server as  
```
-- [DEBUG] RealSimpleHandler::get /
127.0.0.1 - - [22/May/2017 12:18:16] "GET / HTTP/1.1" 200 -
-- [DEBUG] Replying: {u'foo': u'baz', u'config': u'baz'}
```

When I do the full recommended osqueryd enrollment I also get success (from localhost). The command I am running is 
```
[test@localhost osquery]$ osqueryd --verbose \
     --pidfile /tmp/osqueryd.pid \
     --database_path /tmp/osquery.db/ \
     --tls_hostname localhost:8080 \
     --tls_server_certs ./tools/tests/test_server_ca.pem \
     --config_plugin tls \
     --config_tls_endpoint /config \
     --logger_tls_endpoint /logger \
     --logger_plugin tls  \
     --enroll_tls_endpoint /enroll \
     --enroll_secret_path ./tools/tests/test_enroll_secret.txt
```

The output as a result is (with some issues but nothing causing it to fail)
```
I0522 12:31:20.028614 49037 init.cpp:383] osquery initialized [version=2.4.2]
I0522 12:31:20.038682 49037 system.cpp:309] Found stale process for osqueryd (39393)
I0522 12:31:20.038789 49037 system.cpp:344] Writing osqueryd pid (49037) to /tmp/osqueryd.pid
I0522 12:31:20.038909 49037 extensions.cpp:295] Could not autoload extensions: Failed reading: /etc/osquery/extensions.load
I0522 12:31:20.041906 49038 watcher.cpp:465] osqueryd watcher (49037) executing worker (49039)
I0522 12:31:20.051198 49039 init.cpp:380] osquery worker initialized [watcher=49037]
I0522 12:31:20.051337 49039 extensions.cpp:308] Could not autoload modules: Failed reading: /etc/osquery/modules.load
I0522 12:31:20.051416 49039 rocksdb.cpp:205] Opening RocksDB handle: /tmp/osquery.db/
I0522 12:31:20.059790 49039 tls.cpp:68] TLSEnrollPlugin requesting a node enroll key from: https://localhost:8080/enroll
I0522 12:31:20.060077 49039 tls.cpp:205] TLS/HTTPS POST request to URI: https://localhost:8080/enroll
I0522 12:31:20.068608 49039 tls.cpp:205] TLS/HTTPS POST request to URI: https://localhost:8080/config
```
This shows that it has enrolled successfully. However when I try that from a different machine, I'm getting connection denied and the http server isn't even seeing the requests. I ran the same command as above, accounting for IP address, on a different machine that also had the cloned repo and access to the _exact_ same certs but got the following result.

```
I0522 11:39:06.352113 2081939456 init.cpp:380] osquery initialized [version=2.3.3]
I0522 11:39:06.370162 2081939456 system.cpp:262] Found stale process for osqueryd (3258)
I0522 11:39:06.370278 2081939456 system.cpp:297] Writing osqueryd pid (3264) to /tmp/osqueryd.pid
I0522 11:39:06.370535 2081939456 extensions.cpp:292] Could not autoload extensions: Failed reading: /var/osquery/extensions.load
I0522 11:39:06.371306 1064960 watcher.cpp:465] osqueryd watcher (3264) executing worker (3265)
I0522 11:39:06.381777 2081939456 init.cpp:377] osquery worker initialized [watcher=3264]
I0522 11:39:06.382639 2081939456 extensions.cpp:305] Could not autoload modules: Failed reading: /var/osquery/modules.load
I0522 11:39:06.382712 2081939456 rocksdb.cpp:202] Opening RocksDB handle: /tmp/osquery.db/
I0522 11:39:06.538311 2081939456 tls.cpp:68] TLSEnrollPlugin requesting a node enroll key from: https://172.16.83.145:8080/enroll
I0522 11:39:06.538722 2081939456 tls.cpp:205] TLS/HTTPS POST request to URI: https://172.16.83.145:8080/enroll
W0522 11:39:06.542455 2081939456 tls.cpp:75] Failed enrollment request to https://172.16.83.145:8080/enroll (Request error: Connection refused) retrying...
I0522 11:39:06.563887 2138112 interface.cpp:319] Extension manager service starting: /var/osquery/osquery.em
I0522 11:39:07.544639 2081939456 tls.cpp:205] TLS/HTTPS POST request to URI: https://172.16.83.145:8080/enroll
W0522 11:39:07.547210 2081939456 tls.cpp:75] Failed enrollment request to https://172.16.83.145:8080/enroll (Request error: Connection refused) retrying...
^CI0522 11:39:09.376039 2081939456 tls.cpp:205] TLS/HTTPS POST request to URI: https://172.16.83.145:8080/enroll
I0522 11:39:09.378805 2081939456 tls.cpp:68] TLSEnrollPlugin requesting a node enroll key from: https://172.16.83.145:8080/enroll
I0522 11:39:09.378968 2081939456 tls.cpp:205] TLS/HTTPS POST request to URI: https://172.16.83.145:8080/enroll
W0522 11:39:09.380475 2081939456 tls.cpp:75] Failed enrollment request to https://172.16.83.145:8080/enroll (Request error: Connection refused) retrying...
I0522 11:39:10.385767 2081939456 tls.cpp:205] TLS/HTTPS POST request to URI: https://172.16.83.145:8080/enroll
W0522 11:39:10.387958 2081939456 tls.cpp:75] Failed enrollment request to https://172.16.83.145:8080/enroll (Request error: Connection refused) retrying...
```

From the looks of the error at `tls.cpp:75` it looks to be a client certificate issue, but I don't know what to do about that. I have tried supplying the client cert, but don't know if that's right or if I'm doing it properly. 

I even tried simply  curling the URL in the same manner I did from localhost but still get connection refused, with no indication that the server saw it. 

```$ curl https://172.16.83.145:8080 --insecure
curl: (7) Failed to connect to 172.16.83.145 port 8080: Connection refused```

Any help would be much appreciated.

**Comment by spasam** [comment]
@zcatbear Did you check if there is a firewall blocking the connection?

**Comment by spasam** [comment]
BTW, you might get more help on slack: https://osquery.slack.com

**Comment by zcatbear** [comment]
Sorry. I made sure opened the firewall port prior to testing the connection. Exact commands are
```
firewall-cmd --add-port="8080/tcp" --permanent
firewall-cmd --reload
```

**Comment by spasam** [comment]
What do you get when you run the following from osquery agent host:
`openssl s_client -connect 172.16.83.145:8080 -tls1_2`

**Comment by zcatbear** [comment]
```$ openssl s_client -connect 172.16.83.145:8080 -tls1_2
connect: Connection refused
connect:errno=61```

**Comment by spasam** [comment]
I think that implies network error or connection getting dropped. Is your server listening on loopback interface? Run the following on server machine:
`netstat -ntl`

**Comment by zcatbear** [comment]
```$ sudo netstat -ntl
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp        0      0 127.0.0.1:8080          0.0.0.0:*               LISTEN  ```
There was more, but I truncated to only what I needed.

**Comment by spasam** [comment]
Local Address should be 0.0.0.0:8080 or 172.16.83.145:8080. Otherwise, remote client cannot reach your server. You need to fix your server's bind address.

**Comment by zcatbear** [comment]
Got the `curl https://172.16.83.145:8080 --insecure` to work, but the enrollment is still failing on anything not localhost.

```
I0522 13:42:51.494393 2081939456 tls.cpp:205] TLS/HTTPS POST request to URI: https://172.16.83.145:8080/enroll
I0522 13:42:51.505543 2081939456 tls.cpp:68] TLSEnrollPlugin requesting a node enroll key from: https://172.16.83.145:8080/enroll
I0522 13:42:51.505671 2081939456 tls.cpp:205] TLS/HTTPS POST request to URI: https://172.16.83.145:8080/enroll
W0522 13:42:51.511095 2081939456 tls.cpp:75] Failed enrollment request to https://172.16.83.145:8080/enroll (Request error: certificate verify failed) retrying...
```

**Comment by zcatbear** [comment]
After running the above `openssl` command again I have a different output from the server.

```CONNECTED(00000003)
depth=0 C = US, ST = California, O = osquery-testing, CN = localhost
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 C = US, ST = California, O = osquery-testing, CN = localhost
verify error:num=21:unable to verify the first certificate
verify return:1
---
Certificate chain
 0 s:/C=US/ST=California/O=osquery-testing/CN=localhost
   i:/C=US/ST=California/O=osquery-testing/CN=osquery-unittests-ca
---
Server certificate
-----BEGIN CERTIFICATE-----
MIICnjCCAgcCAQgwDQYJKoZIhvcNAQEFBQAwWzELMAkGA1UEBhMCVVMxEzARBgNV
BAgTCkNhbGlmb3JuaWExGDAWBgNVBAoTD29zcXVlcnktdGVzdGluZzEdMBsGA1UE
AxMUb3NxdWVyeS11bml0dGVzdHMtY2EwHhcNMTUwNjEzMDkxMjE2WhcNMjUwNjEw
MDkxMjE2WjBQMQswCQYDVQQGEwJVUzETMBEGA1UECBMKQ2FsaWZvcm5pYTEYMBYG
A1UEChMPb3NxdWVyeS10ZXN0aW5nMRIwEAYDVQQDEwlsb2NhbGhvc3QwggEiMA0G
CSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDvL/dz+qSgwx9dj8EdDeeYV4FOORre
z6CShxiEWylDebeJEXHUXE9y7B9K/CpflGparefmtk7ygKaBMoELTRaVV+BPjx8i
PQNouqlHz/TvVDtpo98b8KKpN39TK7ubikga6hi5zyUxh/CZvBHHGYRJtvSygWGb
FMbt6Tkqv19rz1ArY2E6LZdr6GaulpNmnR6UVF/Ss5g0jytLcWiYgdGzL+tnC6lI
sDB252tAJX1uisTP/S6zT+F3UiNQCuex3VQWsUHXdYMemzWpnzfwQu4fkirl4i5Q
BeClygIsgfsE7tHG2y/cRyqnypsyh6USosa7vMEEFYXRNriisQ8/amHhAgMBAAEw
DQYJKoZIhvcNAQEFBQADgYEAUfVwwbK05VfVdRuZ/vwy2mX2PCwLVCIQoK4eYQxB
xscEKbOpgpk8twWWMqnfvXbzR8glKu7gExtasae07s2NMUYf2x/mSZG+SbpYdYdu
6VgZ8DXNmRxo1GfetMiMnqAuS94+G6eIqZmAQGI/j/Feld/Gi5dGaZ/qW1PSDex8
BQU=
-----END CERTIFICATE-----
subject=/C=US/ST=California/O=osquery-testing/CN=localhost
issuer=/C=US/ST=California/O=osquery-testing/CN=osquery-unittests-ca
---
No client certificate CA names sent
Peer signing digest: SHA512
Server Temp Key: ECDH, P-256, 256 bits
---
SSL handshake has read 1333 bytes and written 434 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: 08C786426BF0A0D95FA71BA2E223C9DB846689A59E028001DDE16B0725C6F472
    Session-ID-ctx: 
    Master-Key: 3154B224A6721210FC70EA519A98FCA02A759436C3DA75D423A38FE5CDBE381B0B5242835E062686B566EABCD8D95AC0
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 300 (seconds)
    TLS session ticket:
    0000 - b0 6f 57 89 6a d6 e6 99-c0 53 08 1b be fe d1 96   .oW.j....S......
    0010 - 92 5d ea 52 0e 40 3f 9e-69 91 f5 79 5c 89 08 ef   .].R.@?.i..y\...
    0020 - 22 7e 5c c0 23 62 7a 93-c5 ac e9 a1 27 b3 11 d6   "~\.#bz.....'...
    0030 - 03 cf 97 88 12 e0 20 51-f1 04 d4 ad 9f 58 4c 27   ...... Q.....XL'
    0040 - ea d2 75 66 54 31 99 51-3c 71 6c d6 9a f7 c5 ff   ..ufT1.Q<ql.....
    0050 - aa f8 69 62 e1 81 ea 30-76 7d a4 ec a3 ac 86 f1   ..ib...0v}......
    0060 - 6e aa 2a 59 96 b3 c8 29-53 3a 2d a4 5a 72 eb 0b   n.*Y...)S:-.Zr..
    0070 - e9 de 8d cd 89 07 14 25-d7 14 0d 6e 97 59 42 86   .......%...n.YB.
    0080 - 40 14 19 61 2f ad b4 4f-52 6d 75 8a ba 22 20 d6   @..a/..ORmu.." .
    0090 - bf e4 76 b6 1d 51 5c eb-a0 0a bd 2f 72 48 44 2f   ..v..Q\..../rHD/

    Start Time: 1495488321
    Timeout   : 7200 (sec)
    Verify return code: 21 (unable to verify the first certificate)```

**Comment by jedi22** [comment]
Are they any errors in the output from the test server?

**Comment by zcatbear** [comment]
No. It doesn't seem to be noticed by the server at all. It's almost as if osquery itself is the one stopping it at line 75 of the plugin due to an invalid cert shown: [link](https://github.com/facebook/osquery/blob/master/osquery/remote/transports/tls.cpp#L75 "here"), but I don't see why that would be triggering because I never set those flags in the first place, so they shouldn't be triggering. 
Error message that's pointing me to that point: 
```W0522 18:18:24.610476 2081939456 tls.cpp:75] Failed enrollment request to https://172.16.83.145:8080/enroll (Request error: certificate verify failed) retrying...```

**Comment by spasam** [comment]
What is in test server ca pem? Should be ca certificate, not server cert.

**Comment by spasam** [comment]
If you run `openssl x509 -text -in test_server_ca.pem` you should see:
```
...
Subject: C=US, ST=California, O=osquery-testing, CN=osquery-unittests-ca
...
```

**Comment by zcatbear** [comment]
For my server I'm running:
```
[test@localhost osquery]$ ./tools/tests/test_http_server.py --tls --persist --cert ./tools/tests/test_server.pem --key ./tools/tests/test_server.key --use_enroll_secret --enroll_secret ./tools/tests/test_enroll_secret.txt 8080

```
When running the command above it _is_ giving me the output as described above:
```Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            f4:2d:79:99:ac:d3:88:f3
    Signature Algorithm: sha1WithRSAEncryption
        Issuer: C=US, ST=California, O=osquery-testing, CN=osquery-unittests-ca
        Validity
            Not Before: Jun 13 09:10:41 2015 GMT
            Not After : Jun 10 09:10:41 2025 GMT
        Subject: C=US, ST=California, O=osquery-testing, CN=osquery-unittests-ca```

**Comment by spasam** [comment]
Not on server side. On the OSQuery agent (--tls_server_certs ./tools/tests/test_server_ca.pem)

**Comment by spasam** [comment]
Sorry, didn't realize this was with OSQuery branch test certs. The server certificate in this case uses 'localhost' as hostname. So it won't work from remote hosts because of OpenSSL host name verification. You will have to use server certificate with correct hostname.

**Comment by zcatbear** [comment]
For further explanation to the above answer. OSquery uses openssl to validate that the server it is sending data to has the same hostname as the CN of the certificate so the default certificates from the repo won't work. You need to generate some certs that will match your hostname to the CN of the certificate. Use openssl to do this:
```
openssl genrsa -out key 2048
openssl req -new -x509 -key key -out cert -days 3650 -subj /CN=<FQDN>
```
where `<FQDN>` is either the fully qualified domain name of the server, or in a pinch, the IP will work. 
From there you just need to restart the server collection with:
```
./tools/tests/test_http_server.py --tls --persist --cert <path>/cert --key <path>/key --use_enroll_secret --enroll_secret ./tools/tests/test_enroll_secret.txt 8080
```
You'll need the cert on both the client and the server. You should just copy it over to your client in whatever fashion you want. I used `scp <user>@<fqdn>:path/to/cert ~/cert` and just run:  
```
osqueryd --verbose --pidfile /tmp/osqueryd.pid --database_path /tmp/osquery.db/ --tls_hostname <host>:8080 --tls_server_certs <path>/cert --config_plugin tls --config_tls_endpoint /config --logger_tls_endpoint /logger --logger_plugin tls --enroll_tls_endpoint /enroll --enroll_secret_path ./tools/tests/test_enroll_secret.txt
```
So in short to fix this issue you need to change the server to bind to it's fqdn and give it some CA certs with the CN as the fqdn.
